File: yurt-system/yurt-system-server/src/main/java/com/github/liulus/yurt/system/security/LoginUserService.java
package com.github.liulus.yurt.system.security;

import com.github.liulus.yurt.convention.bean.BeanUtils;
import com.github.liulus.yurt.system.model.entity.Authority;
import com.github.liulus.yurt.system.model.entity.Role;
import com.github.liulus.yurt.system.model.entity.User;
import com.github.liulus.yurt.system.service.AuthorityService;
import com.github.liulus.yurt.system.service.OrganizationService;
import com.github.liulus.yurt.system.service.RoleService;
import com.github.liulus.yurt.system.service.UserService;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.util.CollectionUtils;

import javax.annotation.Resource;
import java.util.Arrays;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

/**
 * @author liulu
 * @version V1.0
 * @since 2020/11/18
 */
public class LoginUserService implements UserDetailsService {


    @Resource
    private UserService userService;

    @Resource
    private OrganizationService organizationService;

    @Resource
    private RoleService roleService;

    @Resource
    private AuthorityService authorityService;

    private static final String[] ADMIN_USER = {"admin", "liulu"};

    @Override
    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {

        User user = userService.findByUsername(username);
        if (user == null) {
            throw new UsernameNotFoundException("用户 " + username + "不存在!");
        }
        // 构建 loginUser
        LoginUser userDetail = BeanUtils.convert(LoginUser.class, user);
        Optional.ofNullable(userDetail.getOrgId())
                .filter(orgId -> orgId != 0L)
                .map(orgId -> organizationService.findById(orgId))
                .ifPresent(org -> {
                    userDetail.setOrgCode(org.getCode());
                    userDetail.setOrgName(org.getFullName());
                    userDetail.setLevelIndex(org.getLevelIndex());
                });

        // 超级管理员拥有所有权限
        if (Arrays.binarySearch(ADMIN_USER, user.getUsername()) >= 0) {
            List<Authority> authorities = authorityService.findAll();
            userDetail.setAuthorities(authorities.stream().map(Authority::getCode).collect(Collectors.toList()));
            return userDetail;
        }

        List<Role> roles = roleService.findByUserId(user.getId());
        if (CollectionUtils.isEmpty(roles)) {
            return userDetail;
        }
        // 查询用户权限
        List<Authority> roleAuthorities = authorityService.findByRoleIds(roles.stream().map(Role::getId).collect(Collectors.toList()));
//        userDetail.setRoles(roles);
        userDetail.setAuthorities(roleAuthorities.stream().map(Authority::getCode).collect(Collectors.toList()));
        return userDetail;
    }
}

File: yurt-system/yurt-system-server/src/main/kotlin/com/github/liulus/yurt/system/security/LoginUserService.kt
package com.github.liulus.yurt.system.security

import com.github.liulus.yurt.convention.bean.BeanUtils
import com.github.liulus.yurt.system.ext.ADMIN_USER
import com.github.liulus.yurt.system.service.AuthorityService
import com.github.liulus.yurt.system.service.DeptService
import com.github.liulus.yurt.system.service.RoleService
import com.github.liulus.yurt.system.service.UserService
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.security.core.userdetails.UsernameNotFoundException
import org.springframework.util.CollectionUtils
import java.util.Arrays
import java.util.Optional
import javax.annotation.Resource

/**
 * @author liulu
 * @version V1.0
 * @since 2020/11/18
 */
class LoginUserService : UserDetailsService {

    @Resource
    private lateinit var userService: UserService

    @Resource
    private lateinit var deptService: DeptService

    @Resource
    private lateinit var roleService: RoleService

    @Resource
    private lateinit var authorityService: AuthorityService

    override fun loadUserByUsername(username: String): UserDetails {
        val user = userService.findByUsername(username)
            ?: throw UsernameNotFoundException("用户 " + username + "不存在!")

        // 构建 loginUser
        val loginUser = BeanUtils.convert(LoginUser::class.java, user)
        Optional.ofNullable(loginUser.orgId)
            .filter { it > 0L }
            .map { deptService.findById(it) }
            .ifPresent {
                loginUser.orgCode = it.code
                loginUser.orgName = it.name
            }

        // 超级管理员拥有所有权限
        if (Arrays.binarySearch(ADMIN_USER, user.username) >= 0) {
            val authorities = authorityService.findAll()
            val auths = authorities.asSequence().map { it.code }
                .map { SimpleGrantedAuthority(it) }.toList()
            loginUser.setAuthorities(auths)
            return loginUser
        }
        val roles = roleService.findByUserId(user.id!!)
        if (CollectionUtils.isEmpty(roles)) {
            return loginUser
        }

        // 查询用户权限
        val roleAuthorities = authorityService.findByRoleIds(roles.map { it.id!! })
        loginUser.roles = roles.map { it.code.toString() }
        val auths = roleAuthorities.asSequence().map { it?.code }
            .map { SimpleGrantedAuthority(it) }.toList()
        loginUser.setAuthorities(auths)
        return loginUser
    }

}